tests/calculations/test_transfer.py

```diff
@@ -0,0 +1,152 @@
+    assert sorted(calc_info.retrieve_list) == sorted(retrieve_list)
+
+    # Now without symlinks
+    instructions = orm.Dict(dict={'retrieve_files': True, 'remote_files': list_of_files})
+    inputs = {'instructions': instructions, 'source_nodes': list_of_nodes, 'metadata': {'computer': aiida_localhost}}
+    calc_info = generate_calc_job(fixture_sandbox, entry_point_name, inputs)
+    assert sorted(calc_info.remote_symlink_list) == sorted(list())
+    assert sorted(calc_info.remote_copy_list) == sorted(copy_list)
+    assert sorted(calc_info.local_copy_list) == sorted(list())
+    assert sorted(calc_info.retrieve_list) == sorted(retrieve_list)
+
+
+@pytest.mark.usefixtures('clear_database_before_test')
+def test_put_transfer(fixture_sandbox, aiida_localhost, generate_calc_job, tmp_path):
+    """Test a default `TransferCalculation`."""
+
+    file1 = tmp_path / 'file1.txt'
+    file1.write_text('file 1 content')
+    folder = tmp_path / 'folder'
+    folder.mkdir()
+    file2 = folder / 'file2.txt'
+    file2.write_text('file 2 content')
+    data_source = orm.FolderData(tree=str(tmp_path))
+
+    entry_point_name = 'core.transfer'
+    list_of_files = [
+        ('data_source', 'file1.txt', 'folder/file1.txt'),
+        ('data_source', 'folder/file2.txt', 'file2.txt'),
+    ]
+    list_of_nodes = {'data_source': data_source}
+    instructions = orm.Dict(dict={'retrieve_files': False, 'local_files': list_of_files})
+    inputs = {'instructions': instructions, 'source_nodes': list_of_nodes, 'metadata': {'computer': aiida_localhost}}
+
+    # Generate calc_info and verify basics
+    calc_info = generate_calc_job(fixture_sandbox, entry_point_name, inputs)
+    assert isinstance(calc_info, datastructures.CalcInfo)
+    assert isinstance(calc_info.codes_info, list)
+    assert len(calc_info.codes_info) == 0
+    assert calc_info.skip_submit
+
+    # Check that the lists were set correctly
+    copy_list = [
+        (data_source.uuid, 'file1.txt', 'folder/file1.txt'),
+        (data_source.uuid, 'folder/file2.txt', 'file2.txt'),
+    ]
+    assert sorted(calc_info.remote_symlink_list) == sorted(list())
+    assert sorted(calc_info.remote_copy_list) == sorted(list())
+    assert sorted(calc_info.local_copy_list) == sorted(copy_list)
+    assert sorted(calc_info.retrieve_list) == sorted(list())
+
+
+def test_integration_transfer(aiida_localhost, tmp_path):
+    """Test a default `TransferCalculation`."""
+    from aiida.calculations.transfer import TransferCalculation
+    from aiida.engine import run
+
+    content_local0 = 'Content of local file'
+    srcfile_local0 = tmp_path / 'file_local0.txt'
+    srcfile_local0.write_text(content_local0)
+    srcnode_local0 = orm.FolderData(tree=str(tmp_path))
+
+    content_remote = 'Content of remote file'
+    srcfile_remote = tmp_path / 'file_remote.txt'
+    srcfile_remote.write_text(content_remote)
+    srcnode_remote = orm.RemoteData(computer=aiida_localhost, remote_path=str(tmp_path))
```

Comment:
Note that we are testing the mechanism here where both the local and remote machine are the localhost. I am thinking whether this is ok in terms of whether there is now code that is untested when there are two different computers. I think having just this single test case should be fine, but it would be easy to parametrize this test over two different computers: one that is localhost, and a different computer (still localhost but different `Computer` instance) but then with `SshTransport`. You could update the `aiida_localhost` fixture to make the transport configurable

Reply:
>  I am thinking whether this is ok in terms of whether there is now code that is untested when there are two different computers.

I'm not sure what you mean by this; other than when the transfer is checking that all remote folders belong to the same computer, when does the code care that the `localhost` computer is the same or not as the one where the aiida database is deployed? The engine will just treat all target machines the same, it does not make a difference wether the machine is actually the localhost, no?

I mean, I could add this test, I just don't really understand what difference we are testing for (other than perhaps an integrated test between the engine and the ssh transport).

Reply:
I am not sure either, as I said in my comment I was thinking whether there could be some code that is dependent on the transport being used, that is not itself transport code, that could not be tested now. But since I cannot think of it now and you can't either, fine to leave as is